From: Joan Torres <joan.torres@suse.com>
Date: Thu, 4 Apr 2024 11:55:50 +0200
Subject: daemon-handover: Don't logout on handover fail

There can be a situation, where the handover service is stopped and
after a while started again. In that case, the RDP client is
disconnected and when the daemon-handover is started again, a handover
will be started and that can fail.

Do not logout on that failure case. Instead, let the daemon-handover
wait for a future handover.

Bug: https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/issues/193
Bug-Ubuntu: https://launchpad.net/bugs/2072130
Origin: https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/merge_requests/249
---
 src/grd-daemon-handover.c | 26 --------------------------
 src/grd-daemon-system.c   |  1 -
 2 files changed, 27 deletions(-)

diff --git a/src/grd-daemon-handover.c b/src/grd-daemon-handover.c
index 88d7809..e80b1a5 100644
--- a/src/grd-daemon-handover.c
+++ b/src/grd-daemon-handover.c
@@ -38,8 +38,6 @@
 #include "grd-session-rdp.h"
 #include "grd-settings.h"
 
-#define MAX_HANDOVER_WAIT_TIME_MS (20 * 1000)
-
 struct _GrdDaemonHandover
 {
   GrdDaemon parent;
@@ -54,8 +52,6 @@ struct _GrdDaemonHandover
   GCancellable *prompt_cancellable;
 
   unsigned int gnome_remote_desktop_watch_name_id;
-
-  unsigned int logout_source_id;
 };
 
 G_DEFINE_TYPE (GrdDaemonHandover, grd_daemon_handover, GRD_TYPE_DAEMON)
@@ -237,19 +233,6 @@ on_start_handover_finished (GObject      *object,
                 NULL);
 }
 
-static gboolean
-logout (gpointer user_data)
-{
-  GrdDaemonHandover *daemon_handover = user_data;
-
-  g_warning ("[DaemonHandover] Logging out, handover timeout reached");
-
-  daemon_handover->logout_source_id = 0;
-  grd_session_manager_call_logout_sync ();
-
-  return G_SOURCE_REMOVE;
-}
-
 static void
 start_handover (GrdDaemonHandover *daemon_handover,
                 const char        *username,
@@ -259,8 +242,6 @@ start_handover (GrdDaemonHandover *daemon_handover,
     grd_daemon_get_cancellable (GRD_DAEMON (daemon_handover));
   const char *object_path;
 
-  g_assert (!daemon_handover->logout_source_id);
-
   object_path = g_dbus_proxy_get_object_path (
                   G_DBUS_PROXY (daemon_handover->remote_desktop_handover));
 
@@ -273,9 +254,6 @@ start_handover (GrdDaemonHandover *daemon_handover,
     cancellable,
     on_start_handover_finished,
     daemon_handover);
-
-  daemon_handover->logout_source_id =
-    g_timeout_add (MAX_HANDOVER_WAIT_TIME_MS, logout, daemon_handover);
 }
 
 static void
@@ -387,8 +365,6 @@ on_incoming_new_connection (GrdRdpServer      *rdp_server,
 
   if (daemon_handover->use_system_credentials && grd_is_remote_login ())
     show_insecure_connection_prompt (daemon_handover);
-
-  g_clear_handle_id (&daemon_handover->logout_source_id, g_source_remove);
 }
 
 static void
@@ -674,8 +650,6 @@ grd_daemon_handover_shutdown (GApplication *app)
   g_clear_handle_id (&daemon_handover->gnome_remote_desktop_watch_name_id,
                      g_bus_unwatch_name);
 
-  g_clear_handle_id (&daemon_handover->logout_source_id, g_source_remove);
-
   if (daemon_handover->prompt_cancellable)
     {
       g_cancellable_cancel (daemon_handover->prompt_cancellable);
diff --git a/src/grd-daemon-system.c b/src/grd-daemon-system.c
index bab1cea..3d3ab65 100644
--- a/src/grd-daemon-system.c
+++ b/src/grd-daemon-system.c
@@ -394,7 +394,6 @@ on_handle_start_handover (GrdDBusRemoteDesktopRdpHandover *interface,
   return G_DBUS_METHOD_INVOCATION_HANDLED;
 
 err:
-  abort_handover (remote_client);
   g_dbus_method_invocation_return_error (invocation,
                                          G_DBUS_ERROR,
                                          G_DBUS_ERROR_UNKNOWN_OBJECT,
